### PR TITLE
BSM-145: Configure the Business Service attributes using the Admin UI

### PR DESCRIPTION
**Pull request by christianpape**
Jira: http://issues.opennms.org/browse/BSM-145
Bamboo: http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS478

### CODE REVIEW

File: features/vaadin-components/core/src/main/java/org/opennms/netmgt/vaadin/core/KeyValueInputDialogWindow.java [hunk 1]
```diff
@@ -0,0 +1,392 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2015 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/vaadin-components/core/src/main/java/org/opennms/netmgt/vaadin/core/KeyValueInputDialogWindow.java]: should be 2016

File: features/vaadin-components/core/src/main/java/org/opennms/netmgt/vaadin/core/KeyValueInputDialogWindow.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+ * OpenNMS(R) is Copyright (C) 1999-2015 The OpenNMS Group, Inc.
```
Comment (mvrueden) [inline features/vaadin-components/core/src/main/java/org/opennms/netmgt/vaadin/core/KeyValueInputDialogWindow.java]: should be 2016

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java [hunk 3]
```diff
@@ -1,29 +1,29 @@
...
+ *      http://www.gnu.org/licenses/
+ *
  * For more information contact:
- * OpenNMS(R) Licensing <license@opennms.org>
- * http://www.opennms.org/
- * http://www.opennms.com/
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
  *******************************************************************************/
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: :+1:

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java [hunk 4]
```diff
@@ -116,6 +128,24 @@ public BsmAdminPageEditWindow addReductionKeyEdge(String reductionKeyText, Strin
             return this;
         }
 
+        public BsmAdminPageEditWindow addAttribute(String key, String value) throws InterruptedException {
+            newAttributeWindow()
+                    .key(key)
+                    .value(value)
+                    .confirm();
+            wait.until(ExpectedConditions.elementToBeClickable(By.id("addAttributeButton")));
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: This is kinda a duplicate, as we already wait in the confirm() method or is there a particular reason to wait again for another element?

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java [hunk 5]
```diff
@@ -250,6 +280,26 @@ public BsmAdminPageEdgeEditWindow confirm() {
...
+    private class BsmAdminPageAttributeEditWindow {
+        public BsmAdminPageAttributeEditWindow key(String key) throws InterruptedException {
+            findElementById("keyField").clear();
+            findElementById("keyField").sendKeys(key);
+            return this;
+        }
+
+        public BsmAdminPageAttributeEditWindow value(String value) {
+            findElementById("valueField").clear();
+            findElementById("valueField").sendKeys(String.valueOf(value));
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: `value` already is a String, so we do not need `String.valueOf(value)` here.

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java [hunk 6]
```diff
@@ -637,6 +687,34 @@ public void testCanEditPersistedEdge() throws InterruptedException {
...
 
+    @Test
+    public void testCanEditTransientAttribute() throws InterruptedException {
+        // create Business Service with one Attribute
+        final String serviceName = createUniqueBusinessServiceName();
+        BsmAdminPageEditWindow bsmAdminPageEditWindow = bsmAdminPage.openNewDialog(serviceName).addAttribute("foo", "bar");
+        wait.until(pageContainsText("foo=bar"));
+
+        // cancel
+        bsmAdminPageEditWindow.cancel();
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: Please verify afterwards that the attribute `foo = bar` has not been set and please delete the Business Service afterwards.

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java [hunk 7]
```diff
@@ -116,6 +128,24 @@ public BsmAdminPageEditWindow addReductionKeyEdge(String reductionKeyText, Strin
...
+            wait.until(ExpectedConditions.elementToBeClickable(By.id("addAttributeButton")));
+            return this;
+        }
+
+        public BsmAdminPageEditWindow editAttribute(String key, String value) throws InterruptedException {
+            getSelectWebElement("attributeList").selectByVisibleText(key);
+            editAttributeWindow()
+                    .value(value)
+                    .confirm();
+            wait.until(ExpectedConditions.elementToBeClickable(By.id("addAttributeButton")));
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: This is kinda a duplicate, as we already wait in the confirm() method or is there a particular reason to wait again for another element?

File: smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java (hunk 5 above)
Comment (christianpape) [inline smoke-test/src/test/java/org/opennms/smoketest/BSMAdminIT.java]: Sorry, copy & paste mistake

**Comment by mvrueden** [comment]
I tested this and found a couple of things:
 * By default "Key" is set as the default "Key". Let's not do that.
 * If you add an attribute, let's say "Key=Value", and then click `Add Attribute` again, the validator states `String must not be empty` but probably should say something like this: `Key is already defined, please use another name for Key`
 * The value can be empty. Should we enforce an input here? (`input.length() >=1?`)

**Comment by christianpape** [comment]
Fixed the validators and messages but I do not enforce a non-empty input for the attribute's value. In my opinion we should not do this. An empty value has another semantic than an undefined one.

**Comment by j-white** [comment]
I also think that we should allow empty values.

I did notice that the input is focused in the "Value" field after clicking on "Add Attribute". Can we change the default focus to be on the "Key" field instead?

**Pull request merged**